distribution/pull_v2.go

```diff
@@ -218,29 +379,59 @@ func (ld *v2LayerDescriptor) DeltaBase() io.ReadSeeker {
 }
 
 func (ld *v2LayerDescriptor) Close() {
-	if ld.layerDownload != nil {
-		ld.layerDownload.Close()
+	if pullMethod == "resumable" && ld.tmpFile != nil {
+		ld.tmpFile.Close()
+		if err := os.RemoveAll(ld.tmpFile.Name()); err != nil {
+			logrus.Errorf("Failed to remove temp file: %s", ld.tmpFile.Name())
+		}
+	} else {
+		if ld.layerDownload != nil {
+			ld.layerDownload.Close()
+		}
 	}
 }
 
-func (ld *v2LayerDescriptor) Download(ctx context.Context, progressOutput progress.Output) (io.ReadCloser, int64, error) {
-	logrus.Debugf("pulling blob %q", ld.digest)
+func (ld *v2LayerDescriptor) truncateDownloadFile() error {
+	// Need a new hash context since we will be redoing the download
+	ld.verifier = nil
 
-	ld.ctx = ctx
-	ld.layerDownload = nil
-	ld.downloadAttempts = maxDownloadAttempts
-	ld.verifier = ld.digest.Verifier()
+	if _, err := ld.tmpFile.Seek(0, os.SEEK_SET); err != nil {
+		logrus.Errorf("error seeking to beginning of download file: %v", err)
+		return err
+	}
 
-	progress.Update(progressOutput, ld.ID(), "Ready to download")
+	if err := ld.tmpFile.Truncate(0); err != nil {
+		logrus.Errorf("error truncating download file: %v", err)
+		return err
+	}
 
-	return ioutils.NewReadCloserWrapper(ld, func() error { return nil }), ld.src.Size, nil
+	return nil
+}
+
+func (ld *v2LayerDescriptor) Download(ctx context.Context, progressOutput progress.Output) (io.ReadCloser, int64, error) {
+	if os.Getenv("BALENA_ENGINE_PULL_METHOD") == "resumable" {
```

Comment:
we might want to use `BALENA_PULL_METHOD` to keep the balena specific env vars uniform across the code base, I used the simple `BALENA_*` prefix for the storage migration code...